### ISSUE TITLE
Term filters with an array of terms only match on the last term in the array

### ISSUE THREAD

**Issue by pineapplemachine** [elastic/elasticsearch#18719]
**Elasticsearch version**: Reproduced in 1.5.2 and 1.7.2 on 3 June 2016. First noticed around 31 March.

**JVM version**: 1.8.0_40 w/ ES 1.5.2 and 1.7.0_71 w/ ES 1.7.2

**OS version**: OSX 10.11.3 w/ ES 1.5.2 and OSX 10.9.5 w/ ES 1.7.2

**Description of the problem including expected versus actual behavior**:

We have an Elasticsearch index containing documents with ids like `divvybot:1:13`. Previously term filters behaved as expected: The requests described matched and returned one document for every id, since all of the ids in these examples exist in the index I'm querying.

```
curl -XPOST localhost:9200/divvy.bots/divvy.bot/_search -d '
{
    "size": 10,
    "filter": {
        "term": {"_id": [
            "divvybot:1:10", "divvybot:1:11",
            "divvybot:1:12", "divvybot:1:13"
        ]}
    }
}
'
```

Currently the query gets one hit, the document for `divvybot:1:13`.

```
curl -XPOST localhost:9200/divvy.bots/divvy.bot/_search -d '
{
    "size": 10,
    "filter": {
        "term": {"_id": [
            "divvybot:1:10", "divvybot:1:11",
            "divvybot:1:12"
        ]}
    }
}
'
```

This query also gets one hit, in this case the document for `divvybot:1:12`.

In every case I've tried, including for other indexes and other fields, when I attempt to match using a list of terms only the last term in the list is actually matched. I have tried disabling caching for the term filter and enclosing the term within `"bool": {"must": [{ <term> }]}`, as well as within `"query": {"constant_score": { <filter> }}` and found no difference in behavior in any case.

**Steps to reproduce**:
Unknown. Especially confounding is how this issue has occurred simultaneously on two different systems with differing setups, both mine and another developer's.

**Addendum**:

I spent a moment feeling like an idiot for using `term` instead of `terms`, but there is no question that using `term` for an array of terms to match absolutely worked before, and has stopped working without bothering to explain itself.

**Comment by clintongormley** [comment]
>  but there is no question that using term for an array of terms to match absolutely worked before, and has stopped working without bothering to explain itself.

Sorry @pineapplemachine but the `term` filter never worked like this.  In 5.0 we have done a major refactoring of query parsing that will now complain appropriately.

**Comment by pineapplemachine** [comment]
This is a diff from when we added a query using an array of terms given to `term`:

```
    def get_bulk(self, bots):
        """Get information for many bots at once."""
        bot_ids = [bot.resource_id.to_string() for bot in bots]
        BotStorage.logger.error(bot_ids)
        return elasticsearch.helpers.scan(
            self.connection,
            index=BotStorage.index_name, doc_type=BotStorage.doc_type,
            query={'filter': {'term': {'_id': bot_ids}}}
        )
```

I have output from when this was written on 25 May that was as intended - that is, one document was received and processed per id passed to the query. I am and have been using version 1.6.0 Elasticsearch's python API.

It _did_ work, though that it ever did may be the real bug here.

**Comment by clintongormley** [comment]
> It did work, though that it ever did may be the real bug here.

This test:

    PUT test/doc/1?refresh
    {
      "tag": "foo"
    }
    
    PUT test/doc/2?refresh
    {
      "tag": "bar"
    }
    
    GET test/_search
    {
      "query": {
        "constant_score": {
          "filter": {
            "term": {
              "tag": [
                "foo",
                "bar"
              ]
            }
          }
        }
      }
    }

returns a single document:

    {
      "took": 93,
      "timed_out": false,
      "_shards": {
        "total": 5,
        "successful": 5,
        "failed": 0
      },
      "hits": {
        "total": 1,
        "max_score": 1,
        "hits": [
          {
            "_index": "test",
            "_type": "doc",
            "_id": "2",
            "_score": 1,
            "_source": {
              "tag": "bar"
            }
          }
        ]
      }
    }

Tested on 0.20.5, 0.90.0, 1.0.0, 1.1.0, 1.2.1, 1.3.0, 1.4.0, 1.5.0, 1.6.0, 1.7.0.

In 2.0.0, it throws the exception:

    "[term] query does not support array of values"

**Comment by pineapplemachine** [comment]
I checked the commit history and the output we were getting when we added the feature the diff was from. `term` acted as `terms` in 1.5.2 on one system and in 1.7.2 on another starting no later than May 23 - and I recall similarly misusing it to test other features a week or so previous - and abruptly ending around May 31.

**Comment by clintongormley** [comment]
@pineapplemachine I've given you a test showing that this isn't true.  If I've made a mistake somewhere, then please provide a test which proves the opposite.

**Comment by pineapplemachine** [comment]
I tried setting my system clock to May 23, and the test you provided still passed, and the filters that were mistakenly functional before still returned only one document.

Other than traveling back in time, it seems I cannot produce a test because the incorrect behavior was transient. I can only point to our commit history and the results we were getting from our term filter usage until we noticed it was no longer functional around 31 May.
